core/src/test/scala/unit/kafka/admin/ConsumerGroupServiceTest.scala

```diff
@@ -62,6 +62,92 @@ class ConsumerGroupServiceTest {
+        groupState,
+        new Node(1, "localhost", 9092))
+      new DescribeConsumerGroupsResult(Collections.singletonMap(group, KafkaFuture.completedFuture(description)))
+    }
+
+    def offsetsArgMatcherAssignedTopics: util.Map[TopicPartition, OffsetSpec] = {
+      val expectedOffsets = endOffsets.filter{ case (tp, _) => assignedTopicPartitions.contains(tp) }.keySet.map(tp => tp -> OffsetSpec.latest).toMap
+      ArgumentMatchers.argThat[util.Map[TopicPartition, OffsetSpec]] { map =>
+        map.keySet.asScala == expectedOffsets.keySet && map.values.asScala.forall(_.isInstanceOf[OffsetSpec.LatestSpec])
+      }
+    }
+    def offsetsArgMatcherUnassignedTopics: util.Map[TopicPartition, OffsetSpec] = {
+      val expectedOffsets = offsets.asScala.filter{ case (tp, _) => unassignedTopicPartitions.contains(tp) }.keySet.map(tp => tp -> OffsetSpec.latest).toMap
+      ArgumentMatchers.argThat[util.Map[TopicPartition, OffsetSpec]] { map =>
+        map.keySet.asScala == expectedOffsets.keySet && map.values.asScala.forall(_.isInstanceOf[OffsetSpec.LatestSpec])
+      }
+    }
+    when(admin.describeConsumerGroups(ArgumentMatchers.eq(Collections.singletonList(group)), any()))
+      .thenReturn(describeGroupsResult(ConsumerGroupState.STABLE))
+    when(admin.listConsumerGroupOffsets(ArgumentMatchers.eq(group), any()))
+      .thenReturn(AdminClientTestUtils.listConsumerGroupOffsetsResult(offsets))
+    doAnswer(_ => new ListOffsetsResult(endOffsets.asJava)).when(admin).listOffsets(offsetsArgMatcherAssignedTopics, any())
+    doAnswer(_ => new ListOffsetsResult(endOffsets.asJava)).when(admin).listOffsets(offsetsArgMatcherUnassignedTopics, any())
+
+    val (state, assignments) = groupService.collectGroupOffsets(group)
+    assertEquals(Some("Stable"), state)
+    assertTrue(assignments.nonEmpty)
+    // Results should have information for all assigned topic partition (even if there is not Offset's information at all, because they get fills with None)
+    // Results should have information only for unassigned topic partitions if and only if there is information about them (included with null values)
+    assertEquals(assignedTopicPartitions.size + unassignedTopicPartitions.size , assignments.get.size)
+    assignments.map( results =>
+      results.map( partitionAssignmentState =>
+        (partitionAssignmentState.topic, partitionAssignmentState.partition) match {
+          case (Some("testTopic1"), Some(0)) => assertEquals(None, partitionAssignmentState.offset)
+          case (Some("testTopic1"), Some(1)) => assertEquals(Some(100), partitionAssignmentState.offset)
+          case (Some("testTopic1"), Some(2)) => assertEquals(None, partitionAssignmentState.offset)
+          case (Some("testTopic2"), Some(1)) => assertEquals(Some(100), partitionAssignmentState.offset)
+          case (Some("testTopic2"), Some(2)) => assertEquals(None, partitionAssignmentState.offset)
+          case _ => assertTrue(false)
+    }))
```

Comment:
I am not a fan of doing assertions like this. I would rather prefer to use `assertEquals` if possible because the error message is more meaningful in case of failure. We could do something like this:

```
returnedOffsets = assignmentsOpt.map { assignments =>
  assignments.map { assignment => 
    new TopicPartition(....) -> partitionAssignmentState.offset
  }.toMap
}.getOrElse(Map.empty)

expectedOffsets = Map(
   ...
)

assertEquals(expectedOffsets, returnedOffsets)
```

This would replace all the assertions.

Reply:
👍